File: app/src/main/java/com/icerrate/popularmovies/view/movies/detail/MovieDetailFragment.java
package com.icerrate.popularmovies.view.movies.detail;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.AppCompatImageView;
import android.support.v7.widget.AppCompatTextView;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.icerrate.popularmovies.R;
import com.icerrate.popularmovies.data.model.Movie;
import com.icerrate.popularmovies.data.model.Review;
import com.icerrate.popularmovies.data.model.Trailer;
import com.icerrate.popularmovies.utils.InjectionUtils;
import com.icerrate.popularmovies.view.common.BaseFragment;
import com.icerrate.popularmovies.view.common.BaseItemDecoration;
import com.icerrate.popularmovies.view.common.GlideApp;

import java.util.ArrayList;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * @author Ivan Cerrate.
 */

public class MovieDetailFragment extends BaseFragment implements MovieDetailContract.View,
        TrailersAdapter.OnItemClickListener, ReviewsAdapter.OnButtonClickListener {

    public static String KEY_MOVIE = "MOVIE_KEY";

    public static String KEY_TRAILERS = "TRAILERS_KEY";

    public static String KEY_REVIEWS = "REVIEWS_KEY";

    @BindView(R.id.poster)
    public AppCompatImageView posterImageView;

    @BindView(R.id.title)
    public AppCompatTextView titleDateTextView;

    @BindView(R.id.release_date)
    public AppCompatTextView releaseDateTextView;

    @BindView(R.id.rating)
    public AppCompatTextView ratingTextView;

    @BindView(R.id.synopsis)
    public AppCompatTextView synopsisTextView;

    @BindView(R.id.trailers)
    public RecyclerView trailersRecyclerView;

    @BindView(R.id.trailers_no_data)
    public AppCompatTextView trailersNoDataTextView;

    @BindView(R.id.reviews)
    public RecyclerView reviewsRecyclerView;

    @BindView(R.id.reviews_no_data)
    public AppCompatTextView reviewsNoDataTextView;

    private MenuItem shareMenuItem;

    private TrailersAdapter trailersAdapter;

    private ReviewsAdapter reviewsAdapter;

    private MovieDetailPresenter presenter;

    private MovieDetailFragmentListener movieDetailFragmentListener;

    public static MovieDetailFragment newInstance(Movie movie) {
        Bundle bundle = new Bundle();
        if (movie != null) {
            bundle.putParcelable(KEY_MOVIE, movie);
        }
        MovieDetailFragment fragment = new MovieDetailFragment();
        fragment.setArguments(bundle);
        return fragment;
    }

    @Override
    public void onAttach(Activity activity) {
        super.onAttach(activity);
        castOrThrowException(activity);
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        castOrThrowException(context);
    }

    private void castOrThrowException(Context context) {
        try {
            movieDetailFragmentListener = (MovieDetailFragmentListener) context;
        } catch (ClassCastException e) {
            throw new ClassCastException(context.toString() + " must implement MovieDetailFragmentListener");
        }
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        presenter = new MovieDetailPresenter(this,
                InjectionUtils.movieRepository(getContext()));
    }

    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        setHasOptionsMenu(true);
        View view = inflater.inflate(R.layout.fragment_movie_detail, container, false);
        ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        setupView();
        if (savedInstanceState != null) {
            restoreInstanceState(savedInstanceState);
        } else {
            Movie movie = getArguments().getParcelable(KEY_MOVIE);
            presenter.setMovieDetail(movie);
        }
        presenter.loadMovieDetails();
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.menu_movie_detail, menu);
        shareMenuItem = menu.getItem(0);
        presenter.validateMenu();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.share ) {
            presenter.onShareClick();
            return true;
        }
        return false;
    }

    @Override
    protected void saveInstanceState(Bundle outState) {
        outState.putParcelable(KEY_MOVIE, presenter.getMovie());
        outState.putParcelableArrayList(KEY_TRAILERS, presenter.getTrailers());
        outState.putParcelableArrayList(KEY_REVIEWS, presenter.getReviews());
    }

    @Override
    protected void restoreInstanceState(Bundle savedState) {
        Movie movie = savedState.getParcelable(KEY_MOVIE);
        ArrayList<Trailer> trailers = savedState.getParcelableArrayList(KEY_TRAILERS);
        ArrayList<Review> reviews = savedState.getParcelableArrayList(KEY_REVIEWS);
        presenter.loadPresenterState(movie, trailers, reviews);
    }

    private void setupView() {
        //Trailers
        trailersAdapter = new TrailersAdapter(this);
        trailersRecyclerView.setAdapter(trailersAdapter);
        trailersRecyclerView.setLayoutManager(new LinearLayoutManager(getContext(),
                LinearLayoutManager.HORIZONTAL, false));
        trailersRecyclerView.addItemDecoration(new BaseItemDecoration(
                getResources().getDrawable(R.drawable.horizontal_decorator)));
        trailersRecyclerView.setNestedScrollingEnabled(false);
        //Reviews
        reviewsAdapter = new ReviewsAdapter(this);
        reviewsRecyclerView.setAdapter(reviewsAdapter);
        reviewsRecyclerView.setLayoutManager(new LinearLayoutManager(getContext(),
                LinearLayoutManager.VERTICAL, false));
        reviewsRecyclerView.setNestedScrollingEnabled(false);
    }

    @Override
    public void showHeader(String title, String releaseDate, String posterUrl, String backdropUrl, String rating, String synopsis) {
        movieDetailFragmentListener.setBackdropImage(backdropUrl);
        movieDetailFragmentListener.setCollapsingTitle(title);
        movieDetailFragmentListener.setFavoriteOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                presenter.onFavoriteFabClicked();
            }
        });
        titleDateTextView.setText(title);
        GlideApp.with(this)
                .load(posterUrl)
                .placeholder(getResources().getDrawable(R.drawable.poster_placeholder))
                .into(posterImageView);
        releaseDateTextView.setText(releaseDate);
        ratingTextView.setText(getString(R.string.rating, rating));
        synopsisTextView.setText(synopsis);
    }

    @Override
    public void showTrailers(ArrayList<Trailer> trailers) {
        trailersAdapter.addItems(trailers);
    }

    @Override
    public void showTrailersNoData(boolean show) {
        trailersNoDataTextView.setVisibility(show ? View.VISIBLE : View.INVISIBLE);
    }

    @Override
    public void showReviews(ArrayList<Review> reviews) {
        reviewsAdapter.addItems(reviews);
    }

    @Override
    public void showReviewsNoData(boolean show) {
        reviewsNoDataTextView.setVisibility(show ? View.VISIBLE : View.INVISIBLE);
    }

    @Override
    public void showFavoriteState(boolean isFavorite) {
        movieDetailFragmentListener.setFavoriteState(isFavorite);
    }

    @Override
    public void updateFavoriteState(boolean isFavorite) {
        movieDetailFragmentListener.updateFavoriteState(isFavorite);
    }

    @Override
    public void prepareTrailerShare(String trailerUrl) {
        Intent shareIntent = new Intent(Intent.ACTION_SEND);
        shareIntent.setType("text/plain");
        shareIntent.putExtra(Intent.EXTRA_TEXT, trailerUrl);
        startActivity(Intent.createChooser(shareIntent, "Share Trailer link using"));
    }

    @Override
    public void showShareMenu(boolean show) {
        if (shareMenuItem != null) {
            shareMenuItem.setVisible(show);
        }
    }

    @Override
    public void onItemClick(View view) {
        Trailer trailer = (Trailer) view.getTag();
        if (trailer != null) {
            startActivity(new Intent(Intent.ACTION_VIEW,
                    Uri.parse(trailer.getVideoUrl())));
        }
    }

    @Override
    public void onPlayButtonClick(View view) {
        Review review = (Review) view.getTag();
        if (review != null) {
            startActivity(new Intent(Intent.ACTION_VIEW,
                    Uri.parse(review.getUrl())));
        }
    }
}

File: app/src/main/java/com/icerrate/popularmovies/view/movies/detail/MovieDetailFragment.kt
package com.icerrate.popularmovies.view.movies.detail

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.*
import androidx.core.view.MenuProvider
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.core.content.res.ResourcesCompat
import androidx.recyclerview.widget.LinearLayoutManager
import com.bumptech.glide.Glide
import com.icerrate.popularmovies.R
import com.icerrate.popularmovies.data.model.Movie
import com.icerrate.popularmovies.data.model.Review
import com.icerrate.popularmovies.data.model.Trailer
import com.icerrate.popularmovies.databinding.FragmentMovieDetailBinding
import com.icerrate.popularmovies.view.common.BaseFragment
import dagger.hilt.android.AndroidEntryPoint
import com.icerrate.popularmovies.view.common.BaseItemDecoration
import androidx.core.net.toUri
import androidx.core.view.get

/**
 * @author Ivan Cerrate.
 */
@AndroidEntryPoint
class MovieDetailFragment : BaseFragment<FragmentMovieDetailBinding>(
    FragmentMovieDetailBinding::inflate
), TrailersAdapter.OnItemClickListener, ReviewsAdapter.OnButtonClickListener, MenuProvider {

    private var shareMenuItem: MenuItem? = null
    private lateinit var trailersAdapter: TrailersAdapter
    private lateinit var reviewsAdapter: ReviewsAdapter
    private val viewModel: MovieDetailViewModel by viewModels()
    private var movieDetailFragmentListener: MovieDetailFragmentListener? = null


    override fun onAttach(context: Context) {
        super.onAttach(context)
        castOrThrowException(context)
    }

    private fun castOrThrowException(context: Context) {
        try {
            movieDetailFragmentListener = context as MovieDetailFragmentListener
        } catch (_: ClassCastException) {
            throw ClassCastException("$context must implement MovieDetailFragmentListener")
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        requireActivity().addMenuProvider(this, viewLifecycleOwner, Lifecycle.State.RESUMED)
        setupView()
        observeViewModel()
        
        if (savedInstanceState != null) {
            restoreInstanceState(savedInstanceState)
        } else {
            val movie = arguments?.getParcelable<Movie>(KEY_MOVIE)
            movie?.let { viewModel.processIntent(MovieDetailIntent.SetMovieDetail(it)) }
        }
        viewModel.processIntent(MovieDetailIntent.LoadMovieDetails)
    }

    override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
        menuInflater.inflate(R.menu.menu_movie_detail, menu)
        shareMenuItem = menu[0]
        viewModel.processIntent(MovieDetailIntent.ValidateMenu)
    }

    override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
        return if (menuItem.itemId == R.id.share) {
            viewModel.processIntent(MovieDetailIntent.OnShareClick)
            true
        } else {
            false
        }
    }

    override fun saveInstanceState(outState: Bundle) {
        val currentState = viewModel.state.value
        outState.putParcelable(KEY_MOVIE, currentState?.movie)
        outState.putParcelableArrayList(KEY_TRAILERS, currentState?.trailers)
        outState.putParcelableArrayList(KEY_REVIEWS, currentState?.reviews)
    }

    override fun restoreInstanceState(savedState: Bundle) {
        val movie = savedState.getParcelable<Movie>(KEY_MOVIE)
        val trailers = savedState.getParcelableArrayList<Trailer>(KEY_TRAILERS)
        val reviews = savedState.getParcelableArrayList<Review>(KEY_REVIEWS)
        viewModel.processIntent(MovieDetailIntent.RestoreState(movie, trailers, reviews))
    }

    private fun setupView() {
        // Trailers
        trailersAdapter = TrailersAdapter(onItemClickListener = this)
        binding.trailers.adapter = trailersAdapter
        binding.trailers.layoutManager = LinearLayoutManager(
            context,
            LinearLayoutManager.HORIZONTAL,
            false
        )
        binding.trailers.addItemDecoration(
            BaseItemDecoration(ResourcesCompat.getDrawable(resources, R.drawable.horizontal_decorator, resources.newTheme()))
        )
        binding.trailers.isNestedScrollingEnabled = false
        
        // Reviews
        reviewsAdapter = ReviewsAdapter(onButtonClickListener = this)
        binding.reviews.adapter = reviewsAdapter
        binding.reviews.layoutManager = LinearLayoutManager(
            context,
            LinearLayoutManager.VERTICAL,
            false
        )
        binding.reviews.isNestedScrollingEnabled = false
    }

    private fun observeViewModel() {
        viewModel.state.observe(viewLifecycleOwner) { state ->
            handleState(state)
        }
    }

    private fun handleState(state: MovieDetailState) {
        // Handle movie header display
        state.movie?.let { movie ->
            showHeader(movie)
        }
        
        // Handle trailers
        if (state.trailers.isNotEmpty()) {
            trailersAdapter.addItems(state.trailers)
            binding.trailersNoData.visibility = View.INVISIBLE
        } else {
            binding.trailersNoData.visibility = View.VISIBLE
        }
        
        // Handle reviews
        if (state.reviews.isNotEmpty()) {
            reviewsAdapter.addItems(state.reviews)
            binding.reviewsNoData.visibility = View.INVISIBLE
        } else {
            binding.reviewsNoData.visibility = View.VISIBLE
        }
        
        // Handle share menu
        shareMenuItem?.isVisible = state.showShareMenu
        
        // Handle share trailer
        state.shareTrailerUrl?.let { trailerUrl ->
            prepareTrailerShare(trailerUrl)
            viewModel.onShareHandled()
        }
        
        // Handle errors
        state.errorMessage?.let { errorMessage ->
            showError(errorMessage)
            viewModel.onErrorHandled()
        }
    }

    private fun showHeader(movie: Movie) {
        movieDetailFragmentListener?.setBackdropImage(movie.getBackdropUrl(BACKDROP_CODE))
        movieDetailFragmentListener?.setCollapsingTitle(movie.title)
        movieDetailFragmentListener?.setFavoriteOnClickListener {
            viewModel.processIntent(MovieDetailIntent.OnFavoriteFabClicked)
        }
        movieDetailFragmentListener?.setFavoriteState(movie.isFavorite)
        
        binding.title.text = movie.title
        Glide.with(this)
            .load(movie.getPosterUrl(POSTER_CODE))
            .placeholder(resources.getDrawable(R.drawable.poster_placeholder))
            .into(binding.poster)
        binding.releaseDate.text = movie.releaseDate
        val rating = "%.1f".format(movie.voteAverage)
        binding.rating.text = getString(R.string.rating, rating)
        binding.synopsis.text = movie.overview
    }

    private fun prepareTrailerShare(trailerUrl: String) {
        val shareIntent = Intent(Intent.ACTION_SEND).apply {
            type = "text/plain"
            putExtra(Intent.EXTRA_TEXT, trailerUrl)
        }
        startActivity(Intent.createChooser(shareIntent, "Share Trailer link using"))
    }

    override fun onItemClick(view: View) {
        val trailer = view.tag as? Trailer
        trailer?.let {
            startActivity(Intent(Intent.ACTION_VIEW, it.getVideoUrl().toUri()))
        }
    }

    override fun onPlayButtonClick(view: View) {
        val review = view.tag as? Review
        review?.let {
            startActivity(Intent(Intent.ACTION_VIEW, it.url.toUri()))
        }
    }

    companion object {
        const val KEY_MOVIE = "MOVIE_KEY"
        const val KEY_TRAILERS = "TRAILERS_KEY"
        const val KEY_REVIEWS = "REVIEWS_KEY"

        private const val POSTER_CODE = "w500"
        private const val BACKDROP_CODE = "w780"

        fun newInstance(movie: Movie): MovieDetailFragment {
            val bundle = Bundle().apply {
                putParcelable(KEY_MOVIE, movie)
            }
            return MovieDetailFragment().apply {
                arguments = bundle
            }
        }
    }
}